### PR TITLE
Elaborate on FencedFrameConfig

### PR DESCRIPTION
**Pull request by gtanzer**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/pull/67.html" title="Last updated on Apr 17, 2023, 7:17 PM UTC (208b004)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/67/123ccb0...208b004.html" title="Last updated on Apr 17, 2023, 7:17 PM UTC (208b004)">Diff</a>

### CODE REVIEW

**Comment by domfarolino** [comment]
Per our earlier discussion, all that's left here is making all of the optional stuff into the correct "Null or [...]" language, and specifying the visibility type etc right?

**Comment by gtanzer** [comment]
> Per our earlier discussion, all that's left here is making all of the optional stuff into the correct "Null or [...]" language, and specifying the visibility type etc right?

Yes, I believe so.

File: spec.bs
Comment (gtanzer) [reply spec.bs:284]: https://github.com/WICG/fenced-frame/issues/70

File: spec.bs
Comment (domfarolino) [inline spec.bs:525]: I'd make this:

```html
<div algorithm>
  In other to <dfn export>redact a fenced frame config</dfn>, run these steps:
</div>
```

So that it can be referenced as a dfn from other specs.

File: spec.bs
Comment (domfarolino) [inline spec.bs:469]: missing `=` after `urn uuids`

File: spec.bs
Comment (domfarolino) [inline spec.bs:318]: This and the above algorithm should follow the advice in my comment for "redacting" a config.

File: spec.bs
Comment (domfarolino) [inline spec.bs:365]: Will anyone ever directly reference this dfn instead of the two embedder- and content-specific ones? If not, maybe we can just define both types as having the generic values defined below.

Comment (gtanzer) [reply spec.bs:396]: There may be places in the redaction spec where we want to phrase things generally in terms of visibilities. But outside of the FF spec, I don't think so.

File: spec.bs
Comment (gtanzer) [reply spec.bs:556]: I changed it to have `<dfn export>redact a [=fenced frame config=]</dfn>`, lmk if the link inside the definition is a problem.

File: spec.bs
Comment (domfarolino) [inline spec.bs:292]: Do all of these need to be exported (i.e., directly usable by other specifications), or can we reference them indirectly through the algorithms?

File: spec.bs
Comment (domfarolino) [inline spec.bs:303]: same wrt `export`

File: spec.bs
Comment (domfarolino) [inline spec.bs:366]: Can you follow the algorithm trend I did in https://github.com/WICG/fenced-frame/pull/67/commits/321f2a0a10304bba50790c8cc75c064a89a2fa0f with this one, especially regarding usage of infra primitives?

File: spec.bs
Comment (domfarolino) [inline spec.bs:416]: We should be able to link directly via `[FLEDGE](https://wicg.github.io/turtledove)`

File: spec.bs
Comment (domfarolino) [inline spec.bs:679]: Please use `|instance|` syntax inside algorithms for variables here and elsewhere in this PR (as we do in the rest of the spec).

File: spec.bs
Comment (domfarolino) [inline spec.bs:632]: ditto, follow the trend in my previous commit wrt to signature and naming etc.

File: spec.bs
Comment (domfarolino) [inline spec.bs:724]: Please continue the For each iteration map syntax that I used in my previous commit

File: spec.bs
Comment (domfarolino) [inline spec.bs:681]: Should be able to just use `[=DOMString=]` maybe? Or {{DOMString}}, not sure.

File: spec.bs
Comment (domfarolino) [inline spec.bs:684]: Same, you should be able to use {{FenceEvent}}, per https://speced.github.io/bikeshed/#:~:text=idl%2C%20covering%20the%20dfn%20autolinks%20for%20WebIDL%20types%20(the%20%7B%7Bfoo%7D%7D%20shorthand)

File: spec.bs
Comment (domfarolino) [inline spec.bs:726]: Can you replace TODO:ref with the appropriate reference (or if this was an easter-egg for _me_ to fill in let me know and I can help).

File: spec.bs
Comment (domfarolino) [inline spec.bs:676]: Here and elsewhere let's follow the pattern like https://wicg.github.io/turtledove/#:~:text=durationSeconds)%20method%20steps%20are%3A-,If%20this%27s%20relevant%20global%20object%27s%20associated%20Document,-is%20not%20allowed%20to%20use when it comes to accessing the current document.

File: spec.bs
Comment (domfarolino) [inline spec.bs:669]: We should be able to more formally access the traversable navigable via `Let |instance| be [=this=]'s [=relevant global object=]'s [=associated Document=]'s [=Document/node navigable=]'s [=navigable/traversable navigable=]` I think (here and any other places in this PR).

File: spec.bs
Comment (domfarolino) [inline spec.bs:522]: Could we formalize this a bit? I read here that this is a struct-or-null, but in https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/pull/67.html#store-nested-configs we try and iterate over `nested configs` which doesn't really make sense since this is a struct. It sounds like we want this to be:

`nested configs: {
    configs: map<urn, config>;
    visibility: visibility;
}`

in pseudo code? If that's right, let's give this a _member_ that is a map that we can explicitly reference when performing the iteration in the aforementioned link (and anywhere else, if applicable).

**Review by domfarolino — COMMENTED** [review]
Getting close!

File: spec.bs
Comment (gtanzer) [reply spec.bs:292]: I think we can just use the algorithms. Removed the export.

File: spec.bs
Comment (gtanzer) [reply spec.bs:303]: Done.

File: spec.bs
Comment (gtanzer) [reply spec.bs]: Changing the algorithm names to match the models you gave.

File: spec.bs
Comment (gtanzer) [reply spec.bs]: I think we actually don't need visibility to content, like I suspected earlier, but let me finish fleshing out all the fields/redaction and see if that's still the case.

File: spec.bs
Comment (gtanzer) [reply spec.bs]: It gave me an error with node navigable in Document, so I changed it to:

`[=this=]'s [=relevant global object=]'s [=associated Document=]'s [=node navigable=]'s [=navigable/traversable navigable=]`

File: spec.bs
Comment (gtanzer) [reply spec.bs]: Changed to the [=relevant settings object=] to match FLEDGE spec, but lmk if I'm misunderstanding what that means.

File: spec.bs
Comment (domfarolino) [inline spec.bs:337]: nit: also refer to `[=map/size=]` here

Comment (domfarolino) [inline spec.bs:337]: s/meets or exceeds/>=

File: spec.bs
Comment (domfarolino) [inline spec.bs:376]: Indent to make this appear in the loop

File: spec.bs
Comment (domfarolino) [inline spec.bs:395]: This is the piece that's intended to be called during navigation while in parallel, waiting for a pending config right?

File: spec.bs
Comment (domfarolino) [inline spec.bs:425]: Remind me what this will be used for? I'm not sure I understand what the intention is here (especially with the word "lift"). Is this what will be used by APIs that take a config from the "browser process" and generate the appropriate WebIDL object to expose to script?

Comment (domfarolino) [inline spec.bs:425]: Follow the conventions wrt naming and variables in the algorithm signature.

File: spec.bs
Comment (domfarolino) [inline spec.bs:597]: Let's have this algorithm explicitly return "a new fenced frame config instance, with:" _and then_ list the members below.

File: spec.bs
Comment (domfarolino) [inline spec.bs:487]: nit: These three algorithms have an extra space after `<dfn export>`

File: spec.bs
Comment (domfarolino) [inline spec.bs:448]: These two paragraphs are very informal and don't do a lot of linking or describe normative requirements. I'd recommend one of two things:
 1. Make this into a sort of "Introduction" subsection with the `*This section is non-normative*` disclaimer, and then make the following bits with the "preliminary types" into its own section called "Supporting concepts"
 2. Formalize these two paragraphs a bit and add a bunch of linking, and turn the "For example" stuff into its own example or note block.

I slightly prefer (1), but will leave it to you.

File: spec.bs
Comment (domfarolino) [inline spec.bs:550]: I think it would be good to have some sort of note or description about _who_ is responsible for converting a config into a config instance.

File: spec.bs
Comment (domfarolino) [inline spec.bs:452]: Not sure where this should go off the top of my head, but let's also mention somewhere in this section below that the creator of a config is responsible for passing in _everything_ that can go into it, and that there are no optional members, though they can be explicitly set to null. It could be a note, I'm just trying to make the flow clear somewhere. Maybe it would be good in the "Introduction"-style section that I mentioned above that we could add.

File: spec.bs
Comment (gtanzer) [reply spec.bs:425]: I was thinking that we might need an equivalent of RedactedFencedFrameConfig, but I don't think we do anymore. The only complication is that we might need to put the urn uuid back into the config struct.

File: spec.bs
Comment (gtanzer) [reply spec.bs]: Changed this into a different kind of TODO

File: spec.bs
Comment (gtanzer) [reply spec.bs:448]: Made it a non-normative intro.

File: spec.bs
Comment (gtanzer) [reply spec.bs:452]: Added this into the non-normative intro. Though if we are requiring [=fenced frame config=] to be constructed with explicit values for each field, maybe that should be somewhere normative.

File: spec.bs
Comment (domfarolino) [inline spec.bs:761]: This is a little bit strange to me. What does our implementation do? Surely we don't just defer to C++ undefined behavior right? Can we bail out here? (same with the instances below).

File: spec.bs
Comment (domfarolino) [inline spec.bs:483]: Can this just be the number itself? Specs have the concept of a "pointer to foo" basically, but if we could make this just a plain ole number (instead of a pointer to a number elsewhere) that would be easiest. Is there a strong reason not to do this?

File: spec.bs
Comment (domfarolino) [inline spec.bs:756]: I put this as a TODO here that we should follow-up on quickly, but feel free to address this before we land if you want. Basically I think inside [this section](https://wicg.github.io/fenced-frame/#traversable-navigables) we should introduce a new member called "fenced frame config instance" or whatever is similar to what we have on `DocumentLoader` where this can live. That way here and elsewhere we have a concrete thing to initialize, assign, and address wherever we need.

**Review by domfarolino — APPROVED** [review]
Looking great, just a few small comments and I think we're good to go

File: spec.bs
Comment (gtanzer) [inline spec.bs:691]: The reason why I deleted this is: what does it mean for `url` to be "initially null"? All {{FencedFrameConfig}} constructors should specify the value of `url` (which might be null, but it's not like that's a default).

File: spec.bs
Comment (gtanzer) [reply spec.bs:756]: Agree. Where should it live? If we didn't have to deal with urn iframes I think it could just be attached to the fenced content navigable or whatever, but we probably have to attach it to the root.

Comment (gtanzer) [reply spec.bs:756]: That reminds me: this part about fenced frame config mappings is probably wrong for urn iframes (though it is wrong in the same way that our implementation is "wrong"):

```
Each [=traversable navigable=] has a <dfn for="traversable navigable" export>fenced frame config
mapping</dfn>, which is a new [=fenced frame config mapping=].
```

File: spec.bs
Comment (gtanzer) [reply spec.bs:761]: We have a CHECK because this case is currently unreachable. We could make [=mapped url=] not nullable for now, but then we might have to change it later.

File: spec.bs
Comment (gtanzer) [reply spec.bs:483]: It has to be a reference because when you perform a fence-escaping navigation, it subtracts this number from the budget and then sets it to 0 (through the reference, so that it applies to all instances created from the same config).

File: spec.bs
Comment (gtanzer) [reply spec.bs:568]: Is this covered by the non-normative blurb:

```
Each time a <{fencedframe}> navigates to a [=fenced frame config=], it is instantiated as a new
[=fenced frame config instance=], which governs that particular context inside the [=fenced
navigable container/fenced navigable=].
```

or do you mean some additional text right next to this declaration? I imagine the text would be basically identical to that quoted bit.

File: spec.bs
Comment (gtanzer) [reply spec.bs:691]: Oh, you added this back to fix the compile warning...

**Comment by gtanzer** [comment]
@domfarolino Looks fine to merge to me. Still a little stuff to do in followups.

File: spec.bs
Comment (domfarolino) [reply spec.bs:691]: That's right. Maybe I'll just make it a URL that's not initially null, and that has to be provided upon construction.

File: spec.bs
Comment (domfarolino) [reply spec.bs]: I'll make this a spec assert, but if that's really a requirement for now then we should just make the member non-nullable and change that in the future if we ever support that.

File: spec.bs
Comment (domfarolino) [reply spec.bs:568]: No I think this covers it, I must've forgotten about that para during my review.

File: spec.bs
Comment (domfarolino) [reply spec.bs:483]: Where does "the budget" that we subtract _from_ ultimately live? Right now in our spec it seems like a closed loop where the config member is referenced by the instance, and the instance member is referenced by the instantiation algorithm.

File: spec.bs
Comment (domfarolino) [reply spec.bs:755]: So the fenced navigable is indeed a traversable navigable, per https://wicg.github.io/fenced-frame/#fenced-navigable-container-fenced-navigable. That's the root, essentially `content::Page` / `content::FencedFrame` so the member should probably live there, and iframes can climb the tree to reference it. That's basically what the implementation does, right?

Comment (domfarolino) [reply spec.bs:755]: Ah yeah... hmm OK I think for now we should make it a normal member of "navigable" (basically FTN) with some strong Note saying that once URN iframes are gone this will exclusively live on "traversable navigable".

**Pull request merged**
